compiler/lcalc/monomorphize.ml

```diff
@@ -0,0 +1,448 @@
+       (fun _ fields acc ->
+         StructField.Map.fold (fun _ t acc -> collect_typ acc t) fields acc)
+       prg.decl_ctx.ctx_structs acc)
+
+let rec monomorphize_typ
+    (monomorphized_instances : monomorphized_instances)
+    (typ : typ) : typ =
+  match Mark.remove typ with
+  | TStruct _ | TEnum _ | TAny | TClosureEnv | TLit _ -> typ
+  | TArray t1 ->
+    ( TStruct (TypMap.find (Mark.remove t1) monomorphized_instances.arrays).name,
+      Mark.get typ )
+  | TDefault t1 ->
+    TDefault (monomorphize_typ monomorphized_instances t1), Mark.get typ
+  | TArrow (t1s, t2) ->
+    ( TArrow
+        ( List.map (monomorphize_typ monomorphized_instances) t1s,
+          monomorphize_typ monomorphized_instances t2 ),
+      Mark.get typ )
+  | TTuple _ ->
+    ( TStruct (TypMap.find (Mark.remove typ) monomorphized_instances.tuples).name,
+      Mark.get typ )
+  | TOption t1 ->
+    ( TEnum (TypMap.find (Mark.remove t1) monomorphized_instances.options).name,
+      Mark.get typ )
+
+(* We output a typed expr but the types in the output are wrong, it should be
+   untyped and re-typed later. *)
+let rec monomorphize_expr
+    (monomorphized_instances : monomorphized_instances)
+    (e : typed expr) : typed expr boxed =
+  let typ = Expr.maybe_ty (Mark.get e) in
+  match Mark.remove e with
+  | ETuple args ->
+    let new_args = List.map (monomorphize_expr monomorphized_instances) args in
+    let tuple_instance =
+      TypMap.find (Mark.remove typ) monomorphized_instances.tuples
+    in
+    let fields =
+      StructField.Map.of_list
+      @@ List.map2
+           (fun new_arg (tuple_field, _) -> tuple_field, new_arg)
+           new_args tuple_instance.fields
+    in
+    Expr.estruct ~name:tuple_instance.name ~fields (Mark.get e)
+  | ETupleAccess { e = e1; index; _ } ->
+    let tuple_instance =
+      TypMap.find
+        (Mark.remove (Expr.maybe_ty (Mark.get e1)))
+        monomorphized_instances.tuples
+    in
+    let new_e1 = monomorphize_expr monomorphized_instances e1 in
+    Expr.estructaccess ~name:tuple_instance.name
+      ~field:(fst (List.nth tuple_instance.fields index))
+      ~e:new_e1 (Mark.get e)
+  | EMatch { name; e = e1; cases } when EnumName.equal name Expr.option_enum ->
+    let new_e1 = monomorphize_expr monomorphized_instances e1 in
+    let new_cases =
+      EnumConstructor.Map.bindings
+        (EnumConstructor.Map.map
+           (monomorphize_expr monomorphized_instances)
+           cases)
+    in
+    let option_instance =
+      TypMap.find
+        (match Mark.remove (Expr.maybe_ty (Mark.get e1)) with
+        | TOption t -> Mark.remove t
+        | _ -> failwith "should not happen")
+        monomorphized_instances.options
+    in
+    let new_cases =
+      match new_cases with
+      | [(n1, e1); (n2, e2)] -> (
+        match
+          ( Mark.remove (EnumConstructor.get_info n1),
+            Mark.remove (EnumConstructor.get_info n2) )
+        with
+        | "ESome", "ENone" ->
```

Comment:
In this instance yes I can use `EnumConstructor.equal Shared_ast.none_constr` and this should be fixed, but in other places I need to distinguish constructors for monomorphized types and I don't know how to do better than that...

Reply:
Hm so I can see two solutions:
- keep a set of constructors for some/none as part of the ctx
- or share the same some/none construtors among different option types... it might work but that depends on what we need for the different backends

but do we actually need to be able to recognise or manipulate option types specifically once monomorphization has been done ?

Reply:
> but do we actually need to be able to recognise or manipulate option types specifically once monomorphization has been done ?

Yes we need, for the transpilation to C of `HandleDefaultOp` : https://github.com/CatalaLang/catala/blob/c_backend/compiler/scalc/to_c.ml#L509-L517. At some point I need to construct the accumulator variable holding an monomorphized option type, and I need to retrieve the some/none constructors for this monomorphized type.

Concerning the solutions you propose : 

> keep a set of constructors for some/none as part of the ctx

Having monomorphized option types treated as regular enums is really handy though, for computing type dependencies, printing the context, `EMatch`/`EInj` cases, etc. 

> or share the same some/none construtors among different option types... it might work but that depends on what we need for the different backends

Constructor sharing among different enums or field sharing among different structs sounds really scary to me, I'd rather not do this. 

Alternatively, we can add inside `ctx_enums` a polymorphic "metadata" attachment to each of the enum constructors, that is `unit` most of the time except after LCalc monomorphization where it's `kind_of_constr = Some | None | Other`. What do you think ?

Reply:
> Having monomorphized option types treated as regular enums is really handy though, for computing type dependencies, printing the context, EMatch/EInj cases, etc.

ah of course, I meant keeping that set *in addition* to the existing tables (which is somehow similar to your last suggestion)

it's something different but since it can't always be compiled into a function, maybe `HandleDefault*` should not be an operator ?

Reply:
> it's something different but since it can't always be compiled into a function, maybe HandleDefault* should not be an operator ?

You're right, which is why it's not a function anymore, it has its special place as a statement in the `scalc` AST : https://github.com/CatalaLang/catala/blob/c_backend/compiler/scalc/ast.ml#L82